interfaces/ros1/wavemap_ros/src/map_operations/publish_pointcloud_operation.cc

```diff
@@ -0,0 +1,132 @@
+  }
+
+  // Create the message and set the header
+  sensor_msgs::PointCloud obstacle_pointcloud_msg;
+  obstacle_pointcloud_msg.header.frame_id = world_frame_;
+  obstacle_pointcloud_msg.header.stamp = current_time;
+
+  // Define a functor that converts map leaf nodes into pointcloud points
+  auto add_points_for_leaf_node =
+      [min_cell_width = occupancy_map_->getMinCellWidth(),
+       occupancy_threshold = config_.occupancy_threshold_log_odds,
+       &point_msg_vector = obstacle_pointcloud_msg.points](
+          const OctreeIndex& node_index, FloatingPoint node_log_odds) {
+        if (occupancy_threshold < node_log_odds) {
+          if (node_index.height == 0) {
+            const Point3D center = convert::indexToCenterPoint(
+                node_index.position, min_cell_width);
+            point_msg_vector.emplace_back(convert::point3DToPoint32Msg(center));
+          } else {
+            const Index3D node_min_corner =
+                convert::nodeIndexToMinCornerIndex(node_index);
+            const Index3D node_max_corner =
+                convert::nodeIndexToMaxCornerIndex(node_index);
+            for (const Index3D& index :
+                 Grid(node_min_corner, node_max_corner)) {
+              const Point3D center =
+                  convert::indexToCenterPoint(index, min_cell_width);
+              point_msg_vector.emplace_back(
+                  convert::point3DToPoint32Msg(center));
+            }
+          }
+        }
+      };
+
+  // Call the functor for each leaf
+  const Timestamp start_time_internal = Time::now();
+  if (const auto* hashed_wavelet_octree =
+          dynamic_cast<const HashedWaveletOctree*>(occupancy_map_.get());
+      hashed_wavelet_octree) {
+    // For hashed maps, we only process blocks that change
+    hashed_wavelet_octree->forEachBlock(
+        [this, &add_points_for_leaf_node](const Index3D& block_index,
+                                          const auto& block) {
+          const bool block_changed =
+              last_run_timestamp_internal_ < block.getLastUpdatedStamp();
+          if (!config_.only_publish_changed_blocks || block_changed) {
+            block.forEachLeaf(block_index, add_points_for_leaf_node);
+          }
+        });
+  } else if (const auto* hashed_chunked_wavelet_octree =
+                 dynamic_cast<const HashedChunkedWaveletOctree*>(
+                     occupancy_map_.get());
+             hashed_chunked_wavelet_octree) {
+    // For hashed maps, we only process blocks that change
+    hashed_chunked_wavelet_octree->forEachBlock(
+        [this, &add_points_for_leaf_node](const Index3D& block_index,
+                                          const auto& block) {
+          const bool block_changed =
+              last_run_timestamp_internal_ < block.getLastUpdatedStamp();
+          if (!config_.only_publish_changed_blocks || block_changed) {
+            block.forEachLeaf(block_index, add_points_for_leaf_node);
+          }
+        });
```

Comment:
I assume there was no nice way to avoid that code dupplication as only the dynamic cast seems to be different?

Reply:
We can use a generic lambda (i.e. with auto). I updated the code.